processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -0,0 +1,603 @@
+
+// this class contains shared code between GroupByMergingQueryRunnerV2 and GroupByRowProcessor
+public class RowBasedGrouperHelper
+{
+
+  public static Pair<Grouper<RowBasedKey>, Accumulator<Grouper<RowBasedKey>, Row>> createGrouperAccumulatorPair(
+      final GroupByQuery query,
+      final QueryGranularity queryGranularity,
+      final boolean applyExtractionFn,
+      final GroupByQueryConfig config,
+      final ByteBuffer buffer,
+      final int concurrencyHint,
+      final LimitedTemporaryStorage temporaryStorage,
+      final ObjectMapper spillMapper,
+      final AggregatorFactory[] aggregatorFactories
+  )
+  {
+    final GroupByQueryConfig querySpecificConfig = config.withOverrides(query);
+    final Grouper.KeySerdeFactory<RowBasedKey> keySerdeFactory = new RowBasedKeySerdeFactory(
+        query.getDimensions().size(),
+        querySpecificConfig.getMaxMergingDictionarySize() / concurrencyHint
+    );
+    final RowBasedColumnSelectorFactory columnSelectorFactory = new RowBasedColumnSelectorFactory();
+    final Grouper<RowBasedKey> grouper;
+    if (concurrencyHint == 1) {
+      grouper = new SpillingGrouper<>(
+          buffer,
+          keySerdeFactory,
+          columnSelectorFactory,
+          aggregatorFactories,
+          temporaryStorage,
+          spillMapper,
+          querySpecificConfig.getBufferGrouperMaxSize(),
+          querySpecificConfig.getBufferGrouperInitialBuckets()
+      );
+    } else {
+      grouper = new ConcurrentGrouper<>(
+          buffer,
+          concurrencyHint,
+          temporaryStorage,
+          spillMapper,
+          querySpecificConfig.getBufferGrouperMaxSize(),
+          querySpecificConfig.getBufferGrouperInitialBuckets(),
+          keySerdeFactory,
+          columnSelectorFactory,
+          aggregatorFactories
+      );
+    }
+
+    final DimensionSelector[] dimensionSelectors = new DimensionSelector[query.getDimensions().size()];
+    for (int i = 0; i < dimensionSelectors.length; i++) {
+      dimensionSelectors[i] = columnSelectorFactory.makeDimensionSelector(query.getDimensions().get(i));
+    }
+
+    final Accumulator<Grouper<RowBasedKey>, Row> accumulator = new Accumulator<Grouper<RowBasedKey>, Row>()
+    {
+      @Override
+      public Grouper<RowBasedKey> accumulate(
+          final Grouper<RowBasedKey> theGrouper,
+          final Row row
+      )
+      {
+        if (theGrouper == null) {
+          // Pass-through null returns without doing more work.
+          return null;
+        }
+
+        long timestamp = queryGranularity.truncate(row.getTimestampFromEpoch());
+
+        if (queryGranularity instanceof AllGranularity) {
+          timestamp = query.getIntervals().get(0).getStartMillis();
+        }
+
+        columnSelectorFactory.setRow(row);
+        final String[] dimensions = new String[query.getDimensions().size()];
+        for (int i = 0; i < dimensions.length; i++) {
+          final String value;
+          if (applyExtractionFn) {
```

Comment:
That could also potentially control whether to make the agg factories combining factories or not.

raw data -> use the regular factory
partially merged (not raw) data -> use the combining factory